Readmes en context (#688)
kiesraad/abacus: documentatie/gebruikersdocumentatie/verkiezingsproces-en-abacus.md

De UI/UX designs vind je in onze [Figma](https://www.figma.com/design/xHDfsv69Nhmk3IrWC0303B/Public---Kiesraad---Abacus-optelsoftware?node-id=3190-28385&t=VnghjibSJMqrQepm-1).

### Security

Security is een van de belangrijkste pijlers voor Abacus, om een zo hoog mogelijke betrouwbaarheid van de verkiezingsuitslagen te kunnen garanderen. Dit doen we op de volgende manieren:

- Invoer volgens het vier-ogenprincipe: elk proces-verbaal wordt twee keer ingevoerd door verschillende gebruikers.
- Afzenderverificatie: we willen garanderen dat alle bestanden die door Abacus zijn gegenereerd van de juiste afzender komen en dat ze niet bedoeld of onbedoeld gewijzigd zijn. We werken nog aan de methode hiervoor.
- Air-gapped: Abacus wordt air-gapped en lokaal gedraaid.
- We doen interne code-reviews en we laten ook externe code-reviews uitvoeren.
- Ook zal er externe pentesting worden uitgevoerd.
- We controleren telkens of de functionaliteit (nog steeds) aan de wettelijke vereisten voldoet.

### Werkwijze

Ons team werkt volgens deze principes:

- We werken geheel open source op deze openbare GitHub-repository. Ook de pull requests zijn openbaar.

> - We hebben een volledige CI/CD pipeline met tests en linting voor de frontend en backend.

Reviewer: ```suggestion
- We hebben een volledige continuous integration pipeline met tests en linting voor de frontend en backend.
```

We doen geen continuous delivery :)

### Gemeentelijk stembureau, hoofdstembureau, centraal stembureau
Hieronder zie je hoe het gehele proces van het optellen van stemmen verloopt voor landelijke verkiezingen. Andere verkiezingen zijn over het algemeen eenvoudiger.

![Stembureau, GSB, HSB, CSB](/documentatie/gebruikersdocumentatie/img/eml_flow_nl.drawio.png)

- **Blok 2:** Alle processen-verbaal van de stembureaus worden op het gemeentelijk stembureau ingevoerd, waarbij de software de resultaten controleert en eventuele fouten en waarschuwingen weergeeft. Wanneer deze opgelost zijn, kan de invoer worden afgerond en genereert de software het proces-verbaal en het EML_NL-bestand[^2] van het gemeentelijk stembureau.
- **Blok 3:** Het proces van blok 2 herhaalt zich bij de hoofdstembureaus van de kieskringen. Hier worden de processen-verbaal van de gemeentelijke stembureaus ingevoerd, gecontroleerd en afgerond. De software maakt het proces-verbaal en het EML_NL-bestand van het hoofdstembureau.

> - **Blok 3:** Het proces van blok 2 herhaalt zich bij de hoofdstembureaus van de kieskringen. Hier worden de processen-verbaal van de gemeentelijke stembureaus ingevoerd, gecontroleerd en afgerond. De software maakt het proces-verbaal en het EML_NL-bestand van het hoofdstembureau.

Author: Done :)

> - **Blok 4:** Op het centraal stembureau worden de processen-verbaal van de gemeentelijke stembureaus nogmaals ingevoerd ter controle. Tot slot worden de processen-verbaal van de hoofdstembureaus ingevoerd, gecontroleerd en afgerond. Dit leidt tot een EML_NL-bestand en een proces-verbaal van het centraal stembureau. Aan de hand van deze gegevens wordt de zetelverdeling berekend.

Author: geïmplementeerd :)

- De software geeft duidelijk aan hoe onjuiste of inconsistente invoer kan worden opgelost.
- We bieden duidelijke documentatie en instructiemateriaal.

Dit is een voorbeeld van de gebruikersinterface:

![Interface](/documentatie/gebruikersdocumentatie/img/abacus-interface.png)

### Functionaliteit

In de afbeelding hieronder zie je hoe Abacus werkt op het gemeentelijk stembureau. Voor landelijke verkiezingen wordt dit proces herhaald op het hoofdstembureau en centraal stembureau, zoals hierboven beschreven.

**LET OP: dit is een concept en is nog in ontwikkeling.**

![Uitslagvaststelling](/documentatie/gebruikersdocumentatie/img/proces%20uitslagvaststelling.png)

- Bovenaan staat IMPORT: gebruikers kunnen kandidaten en verkiezingsdefinities in EML-formaat importeren in Abacus.
- Links staat INVOER: gebruikers kunnen verkiezingsresultaten invoeren, inclusief verklaringen over ontbrekende stembiljetten/stempassen en klachten van kiezers.
Abacus telt de resultaten op en controleert ze.
- Rechts staat ONDERTEKENING: de data wordt ondertekend met een handtekening van de voorzitter van het gemeentelijk stembureau.

> - Onderaan staat EXPORT: de verkiezingsresultaten worden geëxporteerd als EML_NL-bestand en als PDF-bestand.

Author: 'de server', welke computer is dat?

Voor de architectuur kun je beginnen bij het [Overzicht van de softwarearchitectuur](/documentatie/softwarearchitectuur/Overzicht.md).

De UI/UX designs vind je in onze [Figma](https://www.figma.com/design/xHDfsv69Nhmk3IrWC0303B/Public---Kiesraad---Abacus-optelsoftware?node-id=3190-28385&t=VnghjibSJMqrQepm-1).

### Security

Security is een van de belangrijkste pijlers voor Abacus, om een zo hoog mogelijke betrouwbaarheid van de verkiezingsuitslagen te kunnen garanderen. Dit doen we op de volgende manieren:

- Invoer volgens het vier-ogenprincipe: elk proces-verbaal wordt twee keer ingevoerd door verschillende gebruikers.
- Afzenderverificatie: we willen garanderen dat alle bestanden die door Abacus zijn gegenereerd van de juiste afzender komen en dat ze niet bedoeld of onbedoeld gewijzigd zijn. We werken nog aan de methode hiervoor.
- Air-gapped: Abacus wordt air-gapped en lokaal gedraaid.
- We doen interne code-reviews en we laten ook externe code-reviews uitvoeren.
- Ook zal er externe pentesting worden uitgevoerd.
- We controleren telkens of de functionaliteit (nog steeds) aan de wettelijke vereisten voldoet.

> ### Werkwijze

Author: goeie, ik zal dit even herzien ja!

> ### Werkwijze

Author: Ik heb een aantal keer 'we' verwijderd, in sommige gevallen was het echt mooier om het te laten staan (maar het valt nu niet meer zo op).

## De Kiesraad

De Kiesraad is een onafhankelijk adviesorgaan met 7 leden en staat voor een eerlijk, transparant en controleerbaar verkiezingsproces. De Raad is gevestigd in Den Haag. Het bureau Kiesraad bestaat inmiddels uit meer dan 40 medewerkers.

De taken van de Kiesraad zijn:

- Fungeren als centraal stembureau tijdens landelijke verkiezingen. Dit betekent dat de resultaten uit het hele land worden verzameld en bij elkaar worden opgeteld, waarna de zetelverdeling wordt berekend.
- De overheid en het parlement adviseren over het verkiezingsproces en de Kieswet. Dit doet de Kiesraad gevraagd en ongevraagd.
- Ondersteuning bieden aan gemeentes, politieke partijen en iedereen die actief is in het verkiezingsproces.
- De benodigde hulpmiddelen leveren voor het uitvoeren van verkiezingen, zoals modellen en processen, maar ook technologie en software. Hieronder valt de software voor het optellen van stemmen en de zetelverdeling.

Het wettelijke kader voor de verkiezingen is de Kieswet met het bijbehorende Kiesbesluit en de

> Kiesregeling. Dit kader verandert als gevolg van wetgeving die is ingezet om de ontwikkeling van

Author: Ik heb de links toegevoegd en in een opsomming gezet.

nieuwe verkiezingssoftware te ondersteunen.

Meer informatie over de Kiesraad en de Kieswet vind je op [de website van de Kiesraad](https://www.kiesraad.nl/).

## Het verkiezingsproces

> Zoals iedereen weet stemmen we in Nederland op papier. In het stembureau gelden strikte regels om het stemgeheim te bewaren.

Author: Goed punt! Ik heb in de comment van Oliver een suggestie gedaan :)

![Stemmen in stemhokjes](/documentatie/gebruikersdocumentatie/img/stemhokjes.jpg)

Na het sluiten van de stembussen worden alle stemmen met de hand geteld, waarbij een vier-ogenprincipe wordt gehanteerd. De meeste gemeentes hebben hier een speciaal telsysteem voor.

![Het papieren telproces](/documentatie/gebruikersdocumentatie/img/telproces.jpg)

> </p>

Author: Oeps! Klopt, ik was aan het twijfelen of ik niet toch html wilde gebruiken zodat ik de plaatjes goed kan sizen.

> Wanneer alle stemmen geteld zijn, vult de leider van het stembureau alle uitkomsten in op een proces-verbaal. Het totaal aantal toegelaten kiezers (met stempas, volmacht en kiezerspas), het aantal uitgebrachte stemmen en het eventuele verschil daartussen worden eerst ingevuld. Vervolgens worden alle stemmen per kandidaat ingevuld[^1]. Dat ziet er als volgt uit:

Author: Klopt, ik bedoel de voorzitter. 

```suggestion
Wanneer alle stemmen geteld zijn, vult de voorzitter van het stembureau alle uitkomsten in op een proces-verbaal. Het totaal aantal toegelaten kiezers (met stempas, volmacht en kiezerspas), het aantal uitgebrachte stemmen en het eventuele verschil daartussen worden eerst ingevuld. Vervolgens worden alle stemmen per kandidaat ingevuld[^1]. Dat ziet er als volgt uit:
```

![Een proces-verbaal](/documentatie/gebruikersdocumentatie/img/pv.png)

Dan wordt alles naar een centrale plek in de gemeente gebracht (meestal het gemeentehuis), waar het gemeentelijk stembureau de werkzaamheden overneemt.

[^1]: Het proces waarbij alle stemmen op kandidaatniveau direct in het stembureau worden geteld heet *decentrale stemopneming*. In sommige gemeentes tellen de stembureaus alleen de stemmen op lijstniveau, waarna de stemmen per kandidaat pas bij het gemeentelijk stembureau worden geteld. Dit heet *centrale stemopneming*.

### Gemeentelijk stembureau, hoofdstembureau, centraal stembureau

> Hieronder zie je hoe het gehele proces van het optellen van stemmen verloopt.

Author: ```suggestion
Hieronder zie je hoe het gehele proces van het optellen van stemmen verloopt voor landelijke verkiezingen. Andere verkiezingen zijn over het algemeen eenvoudiger.
```

> Hieronder zie je hoe het gehele proces van het optellen van stemmen verloopt.

Author: ```suggestion
Wanneer alle documenten bij het gemeentelijk stembureau zijn kan het optellen beginnen.

Hieronder zie je hoe het gehele proces van het optellen van stemmen verloopt voor landelijke verkiezingen. Andere verkiezingen zijn over het algemeen eenvoudiger.
```

> Hieronder zie je hoe het gehele proces van het optellen van stemmen verloopt.

Author: Suggestie geïmplementeerd

![Stembureau, GSB, HSB, CSB](/documentatie/gebruikersdocumentatie/img/eml_flow_nl.drawio.png)

- **Blok 1:** Zoals hierboven is uitgelegd, worden de stemmen geteld in het stembureau. Dit leidt tot een proces-verbaal.
- **Blok 2:** Alle processen-verbaal van de stembureaus worden op het gemeentelijk stembureau ingevoerd, waarbij de software de resultaten controleert en eventuele fouten en waarschuwingen weergeeft. Wanneer deze opgelost zijn, kan de invoer worden afgerond en genereert de software een EML_NL-bestand[^2] en een proces-verbaal van het gemeentelijk stembureau.

> - **Blok 3:** Het proces van blok 2 herhaalt zich bij de hoofdstembureaus van de kieskringen. Hier worden de processen-verbaal van de gemeentelijke stembureaus ingevoerd, gecontroleerd en afgerond. Het resultaat is een EML_NL-bestand en een proces-verbaal van het hoofdstembureau.

Author: Ik snap niet helemaal wat je bedoelt met 'readme vereenvoudigen'. Ik denk dat het proces alleen voor het GSB uitleggen te kort is, omdat we uiteindelijk alle verkiezingsvormen gaan doen. In dit document staat in ieder geval wel dat we in de eerste instantie voor GR26 bouwen. Dat kunnen we nog wel beter aangeven in de readme zie ik, bedoel je dat?

> - **Blok 4:** Op het centraal stembureau worden de processen-verbaal van de gemeentelijke stembureaus nogmaals ingevoerd ter controle. Tot slot worden de processen-verbaal van de hoofdstembureaus ingevoerd, gecontroleerd en afgerond. Dit leidt tot een EML_NL-bestand en een proces-verbaal van het centraal stembureau. Aan de hand van deze gegevens wordt de zetelverdeling berekend.

Author: Ja, dat is het doel van de grote zee aan laptops bij de Kiesraad (en de vrijwilligers die we dan hebben).

> - **Blok 4:** Op het centraal stembureau worden de processen-verbaal van de gemeentelijke stembureaus nogmaals ingevoerd ter controle. Tot slot worden de processen-verbaal van de hoofdstembureaus ingevoerd, gecontroleerd en afgerond. Dit leidt tot een EML_NL-bestand en een proces-verbaal van het centraal stembureau. Aan de hand van deze gegevens wordt de zetelverdeling berekend.

Reviewer: Dat geldt alleen voor corrigenda van GSB's, omdat het CSB (de Kiesraad) op dat moment de taken van het HSB heeft overgenomen. De HSB's worden opgeheven na de eerste zitting, dus die hebben geen taak in het verwerken van corrigenda.

> - **Blok 4:** Op het centraal stembureau worden de processen-verbaal van de gemeentelijke stembureaus nogmaals ingevoerd ter controle. Tot slot worden de processen-verbaal van de hoofdstembureaus ingevoerd, gecontroleerd en afgerond. Dit leidt tot een EML_NL-bestand en een proces-verbaal van het centraal stembureau. Aan de hand van deze gegevens wordt de zetelverdeling berekend.

Author: Hmm ja dat is waar. Ik zal de zin verwijderen omdat ie in dit verhaal alleen maar verwarring oplevert.
```suggestion
- **Blok 4:** Op het centraal stembureau worden de processen-verbaal van de hoofdstembureaus ingevoerd, gecontroleerd en afgerond. De software maakt het proces-verbaal en het EML_NL-bestand van het centraal stembureau. Aan de hand van deze gegevens wordt de zetelverdeling berekend.
```

> In deze blokken staat software centraal, en hier komt Abacus dan ook aan bod.

Author: Dit was meer als overgangszinnetje bedoeld naar waar Abacus gebruikt wordt, maar dat klopt inderdaad. 

```suggestion
In deze blokken biedt software ondersteuning, en hier komt Abacus dan ook aan bod.
```

[^2]: EML staat voor Election Markup Language. Het bestandsformaat EML_NL is hierop gebaseerd en aangepast voor Nederlandse verkiezingen.

### Papier is leidend

In het verkiezingsproces is papier leidend. Dat zorgt voor controleerbaarheid en transparantie,
en dit zijn belangrijke waarborgen voor de integriteit van het verkiezingsproces. Het doel is dat alleen de

> uitgebrachte stemmen van invloed zijn op de zetelverdeling en dat dit voor iedereen controleerbaar

Author: Goed punt, dit had ik uit een ander document overgenomen. Ik denk dat deze zin ook weg kan.

is. De software wordt gebruikt ter ondersteuning van het papieren proces.

### Controleprotocol optellingen

> Het controleprotocol optellingen is de afsluiting van het papieren proces. Buiten de software om
> wordt de optelling gecontroleerd, door middels een steekproef een deel met de hand na te rekenen.

Author: ```suggestion
Het controleprotocol optellingen is de afsluiting van het papieren proces. Buiten de software om
wordt de optelling gecontroleerd, door middels een steekproef op drie lijsten een deel met de hand na te rekenen.
```

> wordt de optelling gecontroleerd, door middels een steekproef een deel met de hand na te rekenen.

Reviewer 2: Die "een deel" is nu ambigu. Rekenen we een deel na, nl drie lijsten? Of rekenen we een deel van de drie lijsten na?

> Het controleprotocol optellingen is de afsluiting van het papieren proces. Buiten de software om
> wordt de optelling gecontroleerd, door middels een steekproef een deel met de hand na te rekenen.

Author: ```suggestion
Het controleprotocol optellingen is de afsluiting van het papieren proces. Buiten de software om
wordt de optelling gecontroleerd, door middels een steekproef drie willekeurig gekozen lijsten met de hand na te rekenen.
```

Het doel van Abacus is om de papieren processen-verbaal te digitaliseren en te helpen met de controle op de optellingen, zodat we tel- en schrijffouten kunnen detecteren en voorkomen. De software wordt ontwikkeld ter vervanging van het onderdeel Uitslagvaststelling (U) van OSV2020. Hierbij hebben we een aantal doelen gesteld:

- De integriteit, transparantie en controleerbaarheid van het gehele proces ondersteunen.
- Gebruiksvriendelijke interface en technologie implementeren.
- Volledig open source ontwikkelen.
- De oplossing zelf ontwikkelen en beheren.

Abacus wordt in de eerste instantie ontwikkeld voor de Gemeenteraadsverkiezingen in maart 2026. Meer informatie over de functionaliteit die we gaan bouwen vind je in [Functionaliteit voor Abacus 1.0](/documentatie/functionaliteit/versie-1.0-gr2026.md).

### Gebruiksvriendelijkheid

Omdat het verkiezingsproces snel verloopt en er niet al te vaak verkiezingen zijn, maken de gebruikers slechts korte tijd gebruik van verkiezingssoftware. Daarom is het van groot belang dat Abacus zo gebruiksvriendelijk mogelijk is. Hiervoor implementeren we de volgende maatregelen:

- We doen regelmatige gebruikerstests met echte gebruikers.

> - De invoer wordt niet allemaal op één pagina gedaan zoals in de huidige software, maar in kleinere secties met een validatiecheck na elke invoer.

Reviewer 2: Ik denk dat dit ook werkt zonder verwijzing naar de huidige software:

```suggestion
- De invoer wordt in kleinere secties gedaan met een validatiecheck na elke invoer.
```

- Je kunt de invoer pauzeren en later doorgaan.
- De software wordt geoptimaliseerd voor gebruik met een toetsenbord en is eenvoudig aan te leren.
- De software geeft duidelijk aan hoe onjuiste of inconsistente invoer kan worden opgelost.
- We bieden duidelijke documentatie en instructiemateriaal.

Dit is een voorbeeld van de gebruikersinterface:

![Interface](/documentatie/gebruikersdocumentatie/img/abacus-interface.png)

### Functionaliteit

In de afbeelding hieronder zie je hoe Abacus werkt op het gemeentelijk stembureau. Voor landelijke verkiezingen wordt dit proces herhaald op het hoofdstembureau en centraal stembureau, zoals hierboven beschreven.

**LET OP: dit is een concept en is nog in ontwikkeling.**

> ![Uitslagvaststelling](/documentatie/gebruikersdocumentatie/img/proces%20uitslagvaststelling.png)

Author: Ik zal dat even aan @ring-ring-ring vragen, hij kan dat er vast uit klussen.

> ![Uitslagvaststelling](/documentatie/gebruikersdocumentatie/img/proces%20uitslagvaststelling.png)

Author: Maar is het niet de bedoeling dat we daar in de toekomst geen gebruik meer van maken?

> ![Uitslagvaststelling](/documentatie/gebruikersdocumentatie/img/proces%20uitslagvaststelling.png)

Author: Zoals besproken met @ring-ring-ring laten we dit nog even zoals het is totdat we een andere methode hebben. Eventuele vragen beantwoorden we dan. Dit wordt in het document ook afgevangen doordat we aangeven dat dit nog een concept is dat in ontwikkeling is.

- IMPORT: gebruikers kunnen kandidaten en verkiezingsbestanden in EML-formaat importeren in Abacus.
- INVOER: gebruikers kunnen verkiezingsresultaten invoeren, inclusief verklaringen over ontbrekende stembiljetten/stempassen en klachten van kiezers.
Abacus telt de resultaten op en controleert ze.
De data-analyse leidt tot waarschuwingen voor kleine afwijkingen en foutmeldingen voor grote afwijkingen en inconsistente resultaten.
- ONDERTEKENING: de data wordt ondertekend met een handtekening van de voorzitter van het gemeentelijk stembureau.

> - EXPORT: de verkiezingsresultaten worden geëxporteerd als EML_NL-bestand en als PDF-bestand.

Reviewer 2: Ik moest twee keer kijken voordat ik doorhad dat "IMPORT" etc naar de blauwe kaders rond het Abacus-logo verwijzen. (Had ook niet lang naar de afbeelding gekeken, want "Oh ja, die ken ik.")

Author: Ik heb even toegevoegd waar de woorden in het plaatje staan.

### Technische keuzes en architectuur

> De software wordt gebruikt via de webbrowser, en dit doen we met HTML, CSS en JavaScript. Hiervoor gebruiken we TypeScript, een meer uitgebreide versie van JavaScript met typing. Voor de gebruikersinterface gebruiken we de populaire library React.

Reviewer 2: Ik denk dat we hier kunnen zeggen "dit doen we met HTML, CSS en TypeScript. Voor de gebruikersinterface..."

Mensen die meer willen weten kunnen doorklikken naar het andere document. En daar staat dan ook technisch helemaal correct dat we TypeScript schrijven, maar dat dat als JavaScript in de browser draait.

Reviewer 3: Ergens denk ik ook dat we niet meer uit hoeven te leggen dat een webbrowser intern HTML, CSS en JavaScript gebruikt dus dat dit helemaal weg zou kunnen.

Author: ```suggestion
De software wordt gebruikt via de webbrowser. Hiervoor gebruiken we TypeScript, een meer uitgebreide versie van JavaScript met typing. Voor de gebruikersinterface gebruiken we de populaire library React.
```

> De software wordt gebruikt via de webbrowser, en dit doen we met HTML, CSS en JavaScript. Hiervoor gebruiken we TypeScript, een meer uitgebreide versie van JavaScript met typing. Voor de gebruikersinterface gebruiken we de populaire library React.

Author: Ik heb aan het begin van het kopje een inleidende zin toegevoegd :)

> Voor de backend gebruiken we Rust, een high-level programmeertaal waarmee je efficiënt kunt programmeren. Ook heeft deze taal een 'strict type system' waardoor minder bugs ontstaan. Ook is dit een populaire, open-source programmeertaal met een grote community.

Reviewer 2: Gingen we in "Overwegingen ten aanzien van technische keuzes binnen dit project" niet nog iets toevoegen over de memory safety van Rust? Dat zou ik in beide documenten vermelden.

@praseodym Kun jij een suggestie doen van wat we daar kort maar ook zinvol over willen zeggen?

Reviewer: In dat document staat dat al onder het kopje ["Geheugenveiligheid zonder garbage collector"](https://github.com/kiesraad/abacus/blob/main/documentatie/softwarearchitectuur/overwegingen-talen-en-frameworks.md#geheugenveiligheid-zonder-garbage-collector).

Als we het ook hier willen toevoegen dan bijv.:

```suggestion
Voor de backend gebruiken we Rust, een high-level programmeertaal waarmee je efficiënt kunt programmeren. Het unieke eigendomssysteem van Rust voorkomt geheugenfouten, en de taal heeft een 'strict type system' waardoor minder bugs ontstaan. Ook is dit een populaire, open-source programmeertaal met een grote community.
```

- De integriteit, transparantie en controleerbaarheid van het gehele proces ondersteunen.
- Gebruiksvriendelijke interface en technologie implementeren.
- Volledig open source ontwikkelen.
- De oplossing zelf ontwikkelen en beheren.

Abacus wordt in de eerste instantie ontwikkeld voor de Gemeenteraadsverkiezingen in maart 2026. Meer informatie over de functionaliteit die wordt gebouwd vind je in [Functionaliteit voor Abacus 1.0](/documentatie/functionaliteit/versie-1.0-gr2026.md).

### Gebruiksvriendelijkheid

Omdat er niet al te vaak verkiezingen zijn, maken de gebruikers slechts korte tijd gebruik van verkiezingssoftware. Daarom is het van groot belang dat Abacus zo gebruiksvriendelijk mogelijk is. Daarom proberen we de invoer zo intuitief mogelijk te maken:

- Er worden regelmatige gebruikerstests gedaan met echte gebruikers.
- De invoer wordt in kleinere secties gedaan met een validatiecheck na elke invoer.

> - Je kunt de invoer pauzeren en later doorgaan.

Author: Done :)

![Uitslagvaststelling](/documentatie/gebruikersdocumentatie/img/proces%20uitslagvaststelling.png)

- Bovenaan staat IMPORT: gebruikers kunnen kandidaten en verkiezingsdefinities in EML-formaat importeren in Abacus.
- Links staat INVOER: gebruikers kunnen verkiezingsresultaten invoeren, inclusief verklaringen over ontbrekende stembiljetten/stempassen en klachten van kiezers.
Abacus telt de resultaten op en controleert ze.
- Rechts staat ONDERTEKENING: de data wordt ondertekend met een handtekening van de voorzitter van het gemeentelijk stembureau.
- Onderaan staat EXPORT: de verkiezingsresultaten worden geëxporteerd als EML_NL-bestand en als PDF-bestand.

### Technische keuzes en architectuur

Abacus zal worden gedraaid op een gesloten netwerk zonder internettoegang. De backend draait op één machine en kan via de webbrowser (frontend) worden benaderd vanaf andere computers op het netwerk.

Voor de frontend wordt TypeScript gebruikt, een meer uitgebreide versie van JavaScript met typing. Voor de gebruikersinterface wordt de populaire library React gebruikt.

> De backend wordt gemaakt met Rust, een high-level programmeertaal waarmee je efficiënt kunt programmeren. Het unieke eigendomssysteem van Rust voorkomt geheugenfouten, en de taal heeft een 'strict type system' waardoor minder bugs ontstaan. Ook is dit een populaire, open-source programmeertaal met een grote community.

Reviewer 4: "unieke eigendomssysteem". Is dit een vertaling van een engels begrip? Ik ben redelijk ingevoerd maar snap niet wat hier staat. Kan me voorstellen dat het voor leken nog meer abacadabra is.

Author: Dat klopt, het is een vrij letterlijke vertaling van het concept 'ownership'. Ik heb het teruggezet naar het Engels en een linkje toegevoegd naar de documentatie van Rust voor de mensen die willen weten hoe het werkt. Dit heb ik ook gedaan voor 'strict type system'.

```suggestion
De backend wordt gemaakt met Rust, een high-level programmeertaal waarmee je efficiënt kunt programmeren. Het unieke systeem van '[ownership](https://doc.rust-lang.org/book/ch04-01-what-is-ownership.html)' in Rust voorkomt geheugenfouten, en de taal heeft een '[strict type system](https://doc.rust-lang.org/book/ch03-02-data-types.html)' waardoor minder bugs ontstaan. Ook is dit een populaire, open-source programmeertaal met een grote community.
```

Reviewer 4: Top, prima suggestie. Commit jij 'm?

> Voor de database wordt SQLite gebruikt. Deze library is populair, lichtgewicht en gebruiksvriendelijk, en bovendien is installatie van SQLite niet vereist.

Reviewer 4: - Gebruiksvriendelijk: voor developers bedoelen we hier? Zou dat expliciet maken. 
- We gebruiken SQLite, maar installatie is niet vereist. Misschien iets duidelijker maken wat we hier met installeren bedoelen? Nu lijkt het net alsof we sqlite eerst wel, en dan niet gebruiken.

Author: Gebruiksvriendelijk voor developers is het inderdaad. Ik heb verder een poging gedaan tot uitleggen wat 'installatie is niet vereist' betekent, en daar een linkje naar de site van SQLite bij geplaatst. Ik moet bekennen dat ik dit ook niet volledig snap, maar ik denk dat dit ook info voor de echte kenners is. Die weten wel wat dit betekent.

```suggestion
Voor de database wordt SQLite gebruikt. Deze library is populair, lichtgewicht en gebruiksvriendelijk voor developers, en bovendien is SQLite een '[zero-configuration database](https://www.sqlite.org/zeroconf.html)'. Dit betekent dat installeren niet nodig is, zodat de developer meteen aan de slag kan.
```

Reviewer 4: Prima suggestie, commit jij 'm?

> Security is een van de belangrijkste pijlers voor Abacus, om een zo hoog mogelijke betrouwbaarheid van de verkiezingsuitslagen te kunnen garanderen. Dit doen we op de volgende manieren:

Reviewer 4: ```suggestion
Veiligheid is een van de belangrijkste pijlers voor Abacus, om een zo hoog mogelijke betrouwbaarheid van de verkiezingsuitslagen te kunnen garanderen. Dit doen we op de volgende manieren:
```

- Invoer volgens het vier-ogenprincipe: elk proces-verbaal wordt twee keer ingevoerd door verschillende gebruikers.

> - Afzenderverificatie: hiermee wordt gegarandeerd dat alle bestanden die door Abacus zijn gegenereerd van de juiste afzender komen en dat ze niet bedoeld of onbedoeld gewijzigd zijn. We werken nog aan de methode hiervoor.

Reviewer 4: ```suggestion
- Afzenderverificatie: hiermee wordt gegarandeerd dat bestanden met verkiezingsuitslagen die met Abacus zijn gemaakt, afkomstig zijn van het juiste stembureau en dat ze niet bedoeld of onbedoeld gewijzigd zijn.
```

- Air-gapped: Abacus wordt air-gapped en lokaal gedraaid.
- Er worden interne code-reviews gedaan en we laten ook externe code-reviews uitvoeren.
- Ook zal er externe pentesting worden uitgevoerd.
- Er wordt telkens gecontroleerd of de functionaliteit (nog steeds) aan de wettelijke vereisten voldoet.

### Werkwijze

Ons team werkt volgens deze principes:

- De software is beschikbaar onder de EU PL 1.2-licentie en het team werkt op deze openbare GitHub-repository. Ook de pull requests zijn openbaar.
- We hebben een continuous integration pipeline met tests en linting voor de frontend en backend.

> - Er zijn altijd minimaal 2 reviews en approvals nodig per pull request.

Reviewer 4: ```suggestion
- Als een teamlid code wijzigt, moeten minimaal 2 andere teamleden die wijzigingen reviewen en goedkeuren voordat de wijziging wordt overgenomen.
```

> - Nieuwe code wordt altijd getest. De nieuwe tests moeten in dezelfde PR zitten als de nieuwe functionaliteit.

Reviewer 4: ```suggestion
- Als we nieuwe code toevoegen, maken we meteen tests die de werking van die code controleren. De nieuwe tests moeten in dezelfde PR zitten als de nieuwe functionaliteit.
```
